File: framework/extracted/Ahorn/entities/CodeBreaker.jl
module CodeBreaker

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/CodeBreaker" CodeBreaker(x::Integer, y::Integer, code::String="1234", inputTimeout::Number=5.0)

const placements = Ahorn.PlacementDict(
    "normal" => Ahorn.EntityPlacement(CodeBreaker)
)

function Ahorn.selection(entity::CodeBreaker)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::CodeBreaker, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/codeBreaker.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local codeBreaker = {}

codeBreaker.name = "MaggyHelper/CodeBreaker"
codeBreaker.depth = -100
codeBreaker.fieldInformation = {
    code = {},
    inputTimeout = {
        minimumValue = 0
    }
}
codeBreaker.placements = {
    {
        name = "normal",
        data = {
            code = "1234",
            inputTimeout = 5.0
        }
    }
}

function codeBreaker.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/code_breaker/locked", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    return sprite
end

function codeBreaker.selection(room, entity)
    return utils.rectangle(entity.x - 24, entity.y - 40, 48, 40)
end

return codeBreaker
